### ISSUE TITLE
no-extraneous-dependencies wrongfully raises an error for inner, project-specific modules

### ISSUE THREAD

**Issue by moltar** [benmosher/eslint-plugin-import#1121]
Given the following structure:

```
.
├── app
│   ├── index.js
│   └── node_modules
│       └── module_a
│           └── index.js
├── node_modules
│   └── some-npm-module
├── package-lock.json
└── package.json
```

And `app/index.js` is:

```
const foo = require('module_a')
```

Which is completely legal, and works. But the error is raised anyways. These internal modules are not meant to be added to the `package.json`, because they are project-specific modules.

**Comment by ljharb** [comment]
The point of the rule is that *everything* - with no exceptions - should be listed in package.json.

Being project-specific might mean you'd have nested package.json files, but it has to be in a package.json *somewhere* (i'd recommend the top level).

**Comment by moltar** [comment]
I see and understand. Is there any reason this rule is so strict? I understand not defining npm (external) dependencies, but internal seems odd to include in `package.json`. But I am new to Node.js, so probably not understanding something right. Thanks.

**Comment by ljharb** [comment]
Typically, you'd never want more than one package.json and node_modules per repo. If you have a multi-project repo, you'd want each project to have a single package.json and node_modules. You'd never want a `node_modules` folder that's not a direct sibling of a `package.json`.

Similarly, you'd always want `rm -rf node_modules && npm install` to leave you with the right node_modules (and `node_modules` should always be gitignored).

**Comment by moltar** [comment]
> You'd never want a node_modules folder that's not a direct sibling of a `package.json`.

Please help me understand why not?

The project-specific `app/node_modules/*` modules do not have `package.json` in them at all. It is purely for organization and ability to avoid having to do `require('../../../something')` inside the project.

This setup allows you to just do `require('module_a')` from anywhere within the `./app` dir.

**Comment by ljharb** [comment]
Don't be afraid of the dots :-) The reasons why not include "it can be confusing for developers to understand" - and if you don't have a package.json, then you have no way to track version number ranges (short of committing node_modules into git, which makes git a nightmare to work with).

**Comment by moltar** [comment]
> Don't be afraid of the dots :-)

I know it's a joke, but I feel I still have to address this. I am not afraid of the dots per se, but I do find them to be a nuisance. And I often make a mistake about depth and how many levels to go up.

> The reasons why not include "it can be confusing for developers to understand"

Ok, fair enough, so this is not a purely technical problem. But it's more in a developer experience realm. I was worried that this might have some technical side effects.

> and if you don't have a package.json, then you have no way to track version number ranges (short of committing node_modules into git, which makes git a nightmare to work with).

Well, maybe we are misunderstanding each other. But the inner `node_modules` is directly used in the project. It is not a published or packaged module. So there is no need to have version numbers. It's as-if you placed a code in some other directory, like `./foo/file.js`. You can't version that, right either?

---------------

Here is my thought and mainly the reason why I am researching this and asking these questions. And if it becomes out of the scope of this issue/repo, please let me know!

I feel like this could be a good pattern for a medium-to-large project and can aid refactoring later on. 

All inner modules can be project namespaced, e.g. `./app/node_modules/@project-namespace/${module-name}/`.

Then you are able to just require it without the dots `require('@project-namespace/module')`. And if at any point, you feel like this actually can be extracted into a packaged module, then you have the option to do so. The packaged module will move from `./app/node_modules` into `./node_modules` with proper entry in `package.json`.

Until then, it's just a module, not an `npm`/`packaged` module.

There is actually nothing [in the docs](https://nodejs.org/api/modules.html) that says a Node.js module must have a `package.json` file. 

> In the Node.js module system, each file is treated as a separate module.

Thank you for your thoughts on this!

**Comment by ljharb** [comment]
This plugin already helps ensure that you won’t make mistakes about how many dots; there’s also in-editor projects like importJS that can autocorrect the import paths for you. Another alternative - that i recommend - is to extract as much as possible to separate npm-installed packages.

In that case, you’re using node_modules - something that almost always means “this code was installed by npm” - as an actual development folder, committed to git. This is very confusing.

----

I agree it’s not required by node - that’s why it’s not in the docs - but that’s the way the entire JS community operates, by convention - both developers and tools. Linter rules enforce conventions as well as requirements.

**Comment by moltar** [comment]
Ok, thanks for your input. I will ponder on this more. In meantime, I will close the issue as it is a non-issue. Thank you!
